### PR TITLE
Feature/new ux

### PR DESCRIPTION
**Pull request by nd-novorender**
https://trello.com/c/Qa3UFdBr/821-explorer-new-ux-design

- New UX (behind the `newUx` feature flag)
- New clipping plane HUD interactions
- New UX also includes showing cursor location which required some new functionality

### CODE REVIEW

**Comment by nd-novorender** [comment]
@barabadzhi that's a massive PR. If you have ideas on how to make it more manageable - let me know.
Also we probably should merge i18n PR first, because there might be a lot of conflicts.

**Comment by nd-novorender** [comment]
@barabadzhi branch is synced with `develop`

**Comment by barabadzhi** [comment]
> branch is synced with `develop`
👍🏼 I'll review it as the first thing on my mornings TODO. 🗒️

File: src/features/locationHud/locationHud.tsx [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+    const position = useLastPickSample()?.position;
+    const tmZone = useAppSelector(selectTmZoneForCalc);
+    const [showXyz, setShowXyz] = useState(false);
+    const { t } = useTranslation();
+
+    let content: ReactNode;
+    if (tmZone && !showXyz) {
+        let lon = "-",
+            lat = "-",
+            alt = "-";
```
Comment (barabadzhi) [inline src/features/locationHud/locationHud.tsx:125]: Alternative: `let lon = lat = alt = "-";` 😉

File: src/features/myLocation/useGoToMyLocation.ts [hunk 2]
```diff
@@ -1,35 +1,61 @@
...
                 dispatch(
-                    renderActions.setCamera({
-                        type: cameraType,
-                        goTo: {
-                            position: cameraPos,
-                            rotation: view.renderState.camera.rotation,
-                        },
+                    myLocationActions.setSatus({
+                        status: LocationStatus.Error,
+                        msg: "Your position is outside the scene's boundaries.",
```
Comment (barabadzhi) [inline src/features/myLocation/useGoToMyLocation.ts:58]: Translation?

File: src/features/newDesign/newDesignPopup.tsx [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+                    </Typography>
+                    <Box display="flex" justifyContent="center" mt={2}>
+                        <Button variant="outlined" onClick={accept}>
+                            Got it
+                        </Button>
+                    </Box>
+                </HudPanel>
+            </Box>
+        </>
+    );
```
Comment (barabadzhi) [inline src/features/newDesign/newDesignPopup.tsx:76]: Translation?

File: src/features/quickAccessMenu/clippingMenu.tsx [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+import { useAppDispatch, useAppSelector } from "app/redux-store-interactions";
+import { featuresConfig } from "config/features";
+import { Picker, renderActions, selectPicker } from "features/render";
+
+import { ActiveIcon } from "./activeIcon";
+
+export function ClippingMenu({ onSelect }: { onSelect?: () => void }) {
+    const activePicker = useAppSelector(selectPicker);
+    // const cameraType = useAppSelector(selectCameraType);
+    // const selectingOrthoPoint = activePicker === Picker.OrthoPlane;
```
Comment (barabadzhi) [inline src/features/quickAccessMenu/clippingMenu.tsx:13]: ?

File: src/features/quickAccessMenu/clippingMenu.tsx [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+                }}
+            >
+                <ListItemIcon>
+                    <ActiveIcon
+                        Icon={featuresConfig.clippingPlanes.Icon}
+                        active={activePicker === Picker.CrossSection}
+                    />
+                </ListItemIcon>
+                <ListItemText>Cross section</ListItemText>
+            </MenuItem> */}
```
Comment (barabadzhi) [inline src/features/quickAccessMenu/clippingMenu.tsx:68]: Please, rm or uncomment.

File: src/features/render/renderSlice.ts [hunk 6]
```diff
@@ -796,6 +805,26 @@ function subtreesFromBookmark(
     return subtrees;
 }
 
+const clippingPlaneColors: vec4[] = [
+    [255, 255, 255],
+    [37, 55, 70],
+    // [118, 134, 146],
```
Comment (barabadzhi) [inline src/features/render/renderSlice.ts:811]: ?

File: src/features/render/renderSlice.ts [hunk 7]
```diff
@@ -796,6 +805,26 @@ function subtreesFromBookmark(
...
+    [214, 30, 92],
+    [97, 94, 155],
+    [225, 224, 0],
+    [255, 88, 93],
+].map((v) => [...vec3.scale(v as vec3, v as vec3, 1 / 255), 0.5]);
+
+const clippingPlaneOutlineColors: vec3[] = [
+    [0, 0, 0],
+    [255, 88, 93],
+    // [97, 94, 155],
```
Comment (barabadzhi) [inline src/features/render/renderSlice.ts:821]: ?

File: src/features/widgetGroupPanel/widgetGroupPanel.tsx [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+
+    return (
+        <Box
+            sx={{
+                zIndex: 1050,
+                pointerEvents: "none",
+                position: "absolute",
+                right: 0,
+                bottom: 0,
+                height: open ? `calc(100% - ${theme.spacing(layout.padWidgetsTop ? 9 : 0)})` : undefined,
```
Comment (barabadzhi) [inline src/features/widgetGroupPanel/widgetGroupPanel.tsx:54]: `calc` is one of the slowest CSS features, we can do without here, just calculate percentage directly:
```height: open ? `${100 - theme.spacing(layout.padWidgetsTop ? 9 : 0)}%` : undefined,```

File: src/features/widgetGroupPanel/widgetGroupPanel.tsx [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+            }}
+        >
+            {open && (
+                <HudPanel
+                    sx={{
+                        display: "flex",
+                        flexDirection: "column",
+                        transition: "all 0.2s",
+                        pointerEvents: "auto",
+                        height: `calc(100% - ${theme.spacing(8)})`,
```
Comment (barabadzhi) [inline src/features/widgetGroupPanel/widgetGroupPanel.tsx:64]: same, direct calculation possible

File: src/features/widgetGroupPanel/widgetGroupPanel.tsx [hunk 10]
```diff
@@ -0,0 +1,218 @@
...
+                    open={open}
+                    onClick={() => dispatch(explorerActions.setWidgetGroupPanelState({ ...state, open: !open }))}
+                />
+            </Box>
+        </Box>
+    );
+}
+
+function BottomPanel() {
+    // complicated margins/width/heights and wrappers around the panel and svg are all for good shadows
```
Comment (barabadzhi) [inline src/features/widgetGroupPanel/widgetGroupPanel.tsx:110]: 👍🏼

File: src/features/widgets/widgets.tsx [hunk 11]
```diff
@@ -41,7 +56,167 @@ const Arcgis = lazy(() => import("features/arcgis/arcgis"));
...
+
+        return widgets
+            .map((widget) => {
+                const x = gridSize.width - (widget.x + widget.width);
+                const y = gridSize.height - (widget.y + widget.height);
+                const width = widget.width === gridSize.width ? "100%" : `calc((100% - ${gap}) / ${gridSize.width})`;
+                const left = x === 0 ? "0px" : `calc(((100% - ${gap}) / ${gridSize.width} + ${gap}) * ${x})`;
+                const height =
+                    widget.height === gridSize.height ? "100%" : `calc((100% - ${gap}) / ${gridSize.height})`;
+                const top = y === 0 ? "0px" : `calc(((100% - ${gap}) / ${gridSize.height} + ${gap}) * ${y})`;
```
Comment (barabadzhi) [inline src/features/widgets/widgets.tsx:130]: Can it be simplified omitting `calc` where not absolutely necessary?

File: src/features/widgets/widgets.tsx [hunk 12]
```diff
@@ -41,7 +56,167 @@ const Arcgis = lazy(() => import("features/arcgis/arcgis"));
...
+        }
+    };
+
+    return (
+        <Box
+            sx={{
+                position: "absolute",
+                bottom: 0,
+                right: 0,
+                height: `calc(100% - ${theme.spacing(layout.padWidgetsTop ? 18 : 9)})`,
```
Comment (barabadzhi) [inline src/features/widgets/widgets.tsx:180]: Same

File: src/slices/explorer/slice.ts [hunk 13]
```diff
@@ -35,11 +39,23 @@ const initialState: State = {
     organization: "",
     projectName: "",
     widgets: [],
+    // TODO probably should be stored on the backend instead
```
Comment (barabadzhi) [inline src/slices/explorer/slice.ts:42]: Do we have a task for it (not to forget)?

**Review by barabadzhi — CHANGES_REQUESTED** [review]
I have ignored multiple minor issues, typings, etc.
Those are to be fixed in the future updates to this impl.

Issues from local use:
1. No text on the menu items in the expanded version, only icons.
2. Clicking on a functions group opens "Functions" with group expanded, but
clicking on another group does not expand the relevant group nor it creates the new "Functions" view (if no tool was selected).
3. Not all groups are visible in "Functions". For example, clicking "Search" icon requires to scroll in order to get Search in the view.
4. WISHLIST: Ability to expand "Functions" view similar to the tool view (same button as in the new WidgetHeader).

Skjermbilde 2024-08-27 kl  20 44 11

File: src/features/clippingPlanes/clippingPlaneInteractions.tsx [hunk 14]
```diff
@@ -0,0 +1,417 @@
...
+        let newOffset = originalNormalOffset[3];
+
+        const overlay = document.createElement("div");
+        overlay.style.setProperty("position", "absolute");
+        overlay.style.setProperty("left", "0px");
+        overlay.style.setProperty("top", "0px");
+        overlay.style.setProperty("width", "100%");
+        overlay.style.setProperty("height", "100%");
+        overlay.style.setProperty("z", "10000");
+        document.body.appendChild(overlay);
```
[the remark this answers is not in the document]
Comment (nd-novorender) [reply src/features/clippingPlanes/clippingPlaneInteractions.tsx:119]: I don't think there is any difference between these two approaches

File: src/features/widgetGroupPanel/widgetGroupPanel.tsx (hunk 9 above)
Comment (nd-novorender) [reply src/features/widgetGroupPanel/widgetGroupPanel.tsx:64]: `theme.spacing` is in `px` (final value is `calc(100% - 64px)`) - can it be done without `calc`?

File: src/features/widgets/widgets.tsx (hunk 11 above)
Comment (nd-novorender) [reply src/features/widgets/widgets.tsx:130]: Again all the calcs combine percentages with px and I'm not aware how to do it without `calc`

File: src/features/hud/hud.tsx [hunk 15]
```diff
@@ -1,15 +1,58 @@
...
+
+import GlobalSnackbar from "./globalSnackbar";
 
 export function Hud() {
+    const newDesign = useAppSelector(selectNewDesign);
+
+    return (
+        <>
+            {newDesign ? <HudNew /> : <HudOld />}
+            {/* <NewDesignPopup /> */}
```
[the remark this answers is not in the document]
Comment (nd-novorender) [reply src/features/hud/hud.tsx:23]: Yes - we roll out new design only for Novorender org for now and it's opt-in. New design popup is supposed to be shown when we force new design on people which is going to happen later

File: src/features/locationHud/locationHud.tsx (hunk 1 above)
Comment (nd-novorender) [reply src/features/locationHud/locationHud.tsx:125]: Doesn't work for me :(
image

File: src/features/widgetGroupPanel/widgetGroupPanel.tsx (hunk 8 above)
Comment (nd-novorender) [reply src/features/widgetGroupPanel/widgetGroupPanel.tsx:54]: `theme.spacing` returns `px`

**Review by barabadzhi — APPROVED** [review]
LGTM 👍🏼

File: src/components/widgetHeader.tsx [hunk 16]
```diff
@@ -14,6 +14,7 @@ import {
     selectNewDesign,
     selectPositionedWidgets,
 } from "slices/explorer";
+import { mixpanel } from "utils/mixpanel";
```
Comment (barabadzhi) [inline src/components/widgetHeader.tsx:17]: Unused? 🤔

File: src/features/widgetList/widgetList.tsx [hunk 17]
```diff
@@ -4,6 +4,7 @@ import { useAppDispatch, useAppSelector } from "app/redux-store-interactions";
 import { FeatureGroupKey, featuresConfig, type WidgetKey } from "config/features";
 import GroupedWidgetList from "features/groupedWidgetList/groupedWidgetList";
 import { explorerActions, selectNewDesign, selectWidgets } from "slices/explorer";
+import { mixpanel } from "utils/mixpanel";
```
Comment (barabadzhi) [inline src/features/widgetList/widgetList.tsx:7]: Same?

**Review by barabadzhi — APPROVED** [review]

**Pull request merged**